### PR TITLE
Transition to JSON logging for the Linux Dedicated plan

### PR DESCRIPTION
**Pull request by davidmrdavid**
Previously, logs on Linux Dedicated were interfacing with the legacy regex endpoint. This meant that some of our logs required extra post-processing for them to behave as expected. With this change, we'll be writing to the new JSON endpoint which should _just work_.

_Also, this change will **need** to be backported_

### Issue describing the changes in this PR
N/A

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [x] Otherwise: Backport tracked by issue/PR #1733
* [x] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: test/Common/TestHelpers.cs
Comment (davidmrdavid) [inline test/Common/TestHelpers.cs:260]: Otherwise we get an exception about not being able to read the log-file's contents because another process is using it

**Comment by davidmrdavid** [comment]
I tested this on a linux dedicated instance already. Will test again early next week so we can merge this. I believe this needs to be backported to V1 as well though

**Review by ConnorMcMahon — COMMENTED** [review]
LGTM

File: test/Common/TestHelpers.cs
Comment (davidmrdavid) [reply test/Common/TestHelpers.cs:260]: It's a tricky one :) . Before, I was using `File.ReadAllLines` which apparently attempts to open the file using a restrictive, read-only, file permission. This is incompatible with our file logger's write permissions, so it errors out. With this, we are opening the file using a "read-write" permission so it works just fine :)

Comment (davidmrdavid) [reply test/Common/TestHelpers.cs:260]: I got the description, **and the method itself,** from this SO answer: https://stackoverflow.com/questions/12744725/how-do-i-perform-file-readalllines-on-a-file-that-is-also-open-in-excel

Comment (ConnorMcMahon) [reply test/Common/TestHelpers.cs:260]: Let's directly link to that SO answer in our comment.

**Comment by davidmrdavid** [comment]
Hey @ConnorMcMahon, in my latest commit I refactored the JSON-logging tests to use the `WaitUntilTrue` approach and also simplify some of their shared logic. I'd love to get a final LGTM on it, thanks!

**Comment by davidmrdavid** [comment]
Also, since this needs to be backported to V1, should we wait on the merge until we have a V1 PR? Just to make sure we don't forget about it.

**Comment by ConnorMcMahon** [comment]
For V1, just open a ticket, and reference the ticket in this PR (and mention this PR in the ticket). Then you can check off the sub-box in that category.

File: test/Common/TestHelpers.cs
Comment (davidmrdavid) [reply test/Common/TestHelpers.cs:290]: I didn't see this before, I'll link it

**Review by ConnorMcMahon — APPROVED** [review]
LGTM!

If we have tested this end-to-end before we start the validation step, I am 100% on board for merging this in for v2.4.2.

**Comment by davidmrdavid** [comment]
Thanks! I don't plan to modify the code further, but this could use just a tad more testing so I won't be merging this for 2.4.2 :)

File: test/Common/DurableTaskEndToEndTests.cs
Comment (ConnorMcMahon) [inline test/Common/DurableTaskEndToEndTests.cs:662]: Is EventId the best way of identifying provider? I get that these event ids are unique across providers (at least I believe so), but there should be a Provider field that is more conclusive right?

File: test/Common/DurableTaskEndToEndTests.cs
Comment (ConnorMcMahon) [inline test/Common/DurableTaskEndToEndTests.cs:648]: Is there a reason we prioritized fields like activity ID and related activity ID over other fields like "InstanceId", "Partition", "MessageId", "Message", etc?

Also, it may be slightly more performant (and imo easier to read) if you perform checks one by one.

For instance:

```csharp
var jsonLines = TestHelpers.WriteSafeReadAllLines(LinuxAppServiceLogger.LogginPath).Select(line => JObject.Parse(line));

if (jsonLines.Any(json => !TestHelpers.IsValidJson(json))
{
    return false;
}

if (!jsonLines.Any(json => json.GetValue<string>("ProviderName") == "DurableTask.AzureStorage"))) {
    return false;
}

//etc
```

**Review by ConnorMcMahon — APPROVED** [review]
Nits on tests, but otherwise looks good.

File: test/Common/DurableTaskEndToEndTests.cs
Comment (davidmrdavid) [reply test/Common/DurableTaskEndToEndTests.cs:648]: I think the only reason why I prioritized the ActivityID fields is that, in the past, during development, we had received faulty logs which missed them. I realize all fields are relevant, this test is just particularly informed by a previous bug. To summarize: it was an arbitrary decision.

I also find it surprising that doing a `.Any()` test might be faster than the current approach, since with the current approach we only iterate through the data once. Regardless, since this is a test, I do think readability beats performance, so I'll incorporate your suggestion :)

Comment (ConnorMcMahon) [reply test/Common/DurableTaskEndToEndTests.cs:648]: It's hard to say whether it is more performant or not.

My expectation is right now we are doing ~5x checks over 1 iteration, and we have to complete the entire iteration before returning.

With this approach, while we are iterating 5 times, which introduces overhead, we are only doing 1x check per iteration, so in the case where one of the earlier cases fails, we catch it earlier and don't have to do subsequent iterations/checks. 

.Any() also shortcircuits as soon as any value matches that condition, so if the logs have 50 lines, and all the conditions are met by the first 10 lines, then it should in theory be faster.

Overall, like we said though, this is a test, so perf isn't the top priority (at least not in the order of a few million CPU cycles).

**Comment by davidmrdavid** [comment]
@ConnorMcMahon, only requesting a re-review since I changed the testing code after your last approval. A quick skim should do.

File: src/WebJobs.Extensions.DurableTask/LinuxAppServiceLogger.cs
Comment (ConnorMcMahon) [inline src/WebJobs.Extensions.DurableTask/LinuxAppServiceLogger.cs:29]: Before we merge, we should just 100% verify that this is indeed the file path listened to by linux app services.

**Review by ConnorMcMahon — APPROVED** [review]
LGTM!

File: src/WebJobs.Extensions.DurableTask/LinuxAppServiceLogger.cs
Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/LinuxAppServiceLogger.cs:29]: Confirmed and e2e tested

**Pull request merged**
